specs/vep-2272-complete-data-job-configuration-persistence/README.md

```diff
@@ -0,0 +1,161 @@
+subsequently retrieves the deployment data from Kubernetes. This two-step process introduces performance degradation to
+the Control Service, as it requires additional time and resources to fetch information from multiple sources.
+
+Moreover, in the event of a Kubernetes namespace loss, there is a potential risk of losing critical data job deployment
+configurations. In such scenarios, recovering the lost data jobs becomes a manual and complex restoration process.
+This adds further complexity and potential downtime for users relying on the Control Service for their data job
+management needs.
+
+To address these issues, it is essential to optimize the storage and retrieval of data job deployments. By ensuring that
+all relevant properties are stored consistently, in the database, so we can streamline the process and enhance the
+overall the efficiency of the Control Service APIs.
+
+By improving the efficiency and reliability of data job deployments, we aim to optimize the overall system performance
+and provide a seamless experience for our users.
+
+## Requirements and goals
+
+### Goals
+
+1. Identify the specific data job configuration that is stored only within the Kubernetes cluster.
+2. Transition the primary source of truth for all data jobs configurations from both Kubernetes and database to just the
+   database, in order to centralize data storage, improve data integrity, and enable easier access and manipulation.
+3. Enhance the performance of Control Service APIs by loading the data job configuration from only one data source
+   (Control Service database).
+4. Prepare Control Service for future enhancement of the migration process and ease of restoration by centralizing and
+   storing all data job configurations within the database.
+
+### Non-Goals
+
+1. Implement automatic data jobs restore capability in case of the loss of Kubernetes namespace whereby the Control
+   Service automatically retrieves the configuration from a database and seamlessly restores the data jobs.
+
+## High-level design
+
+These diagrams below show a high-level design view of the changes that would be proposed in this VEP.
+
+|                       Deploy Data Job                       |                           Read Data Job Deployment Configurations                           |
+|:-----------------------------------------------------------:|:-------------------------------------------------------------------------------------------:|
+| ![deploy_data_job_diagram.png](deploy_data_job_diagram.png) | ![read_data_job_deployment_config_diagram.png](read_data_job_deployment_config_diagram.png) |
+
+The proposed design will introduce changes to the Control Service (deployment logic, GraphQL API, etc.), as well as to
+the database model.
+
+When the Control Service receives an API call to deploy data job, it will store the complete deployment configuration in
```

Comment:
You have chosen Write-Through strategy (data is simultaneously updated to db and k8s) . This has clear benefits as read after write consistency. But it's more brittle as potentially would result in slower API and reduces scalability. 

What about employing asynchronous write-behind (data is updated in db and async in k8s) . This is really the kubernetes operator pattern. We have already logic that facilitates this process in DeploymentMonitorSync functionality.

Reply:
Currently, the deployment process is asynchronous - first, the Kubernetes CronJob is created/updated, and then eventually the database flags are updated. This doesn't give us any guarantees that both sources will be in sync.

I am in favor of your idea. We will move the Kubernetes CronJob management logic into a separate async process which will keep the CronJobs in sync with the database. Also, in this way, we will be able to achieve a much easier migration of the data jobs.

Although this logic is more suitable for Kubernetes Operator such an implementation would require much more effort.

Reply:
I will incorporate this approach in the detail design section.